### ISSUE TITLE
pgbouncer crashes if connection is authenticated using “auth_query”, and is then added to the “auth_file”

### ISSUE THREAD

**Issue by gclough** [pgbouncer/pgbouncer#90]
If you’re authenticating via the “auth_query” parameter and then you add the same user into the “auth_file”, then the next time you try to connect pgbouncer crashes.  It appears that it can’t handle the fact that the same user now exists twice, and the `put_in_order()` function triggers the crash:

```
epoll_wait(4, {}, 32, 334)              = 0
stat("/etc/pgbouncer/userlist.txt", {st_mode=S_IFREG|0640, st_size=18, ...}) = 0
epoll_wait(4, {}, 32, 334)              = 0
stat("/etc/pgbouncer/userlist.txt", {st_mode=S_IFREG|0640, st_size=18, ...}) = 0
epoll_wait(4, {}, 32, 334)              = 0
stat("/etc/pgbouncer/userlist.txt", {st_mode=S_IFREG|0640, st_size=18, ...}) = 0
epoll_wait(4, {}, 32, 334)              = 0
stat("/etc/pgbouncer/userlist.txt", {st_mode=S_IFREG|0640, st_size=18, ...}) = 0
epoll_wait(4, {}, 32, 334)              = 0
stat("/etc/pgbouncer/userlist.txt", {st_mode=S_IFREG|0640, st_size=32, ...}) = 0
stat("/etc/pgbouncer/userlist.txt", {st_mode=S_IFREG|0640, st_size=32, ...}) = 0
open("/etc/pgbouncer/userlist.txt", O_RDONLY) = 14
fstat(14, {st_mode=S_IFREG|0640, st_size=32, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ff72bd5f000
read(14, "\"postgres\" \"blah\"\n\"greg\" \"blah\"\n", 4096) = 32
close(14)                               = 0
munmap(0x7ff72bd5f000, 4096)            = 0
write(2, "2015-11-12 18:23:06.169 2182 DEB"..., 842015-11-12 18:23:06.169 2182 DEBUG loading auth_file: "/etc/pgbouncer/userlist.txt"
) = 84
write(3, "2015-11-12 18:23:06.169 2182 DEB"..., 84) = 84
epoll_wait(4, {}, 32, 333)              = 0
stat("/etc/pgbouncer/userlist.txt", {st_mode=S_IFREG|0640, st_size=32, ...}) = 0
epoll_wait(4, {}, 32, 334)              = 0
stat("/etc/pgbouncer/userlist.txt", {st_mode=S_IFREG|0640, st_size=32, ...}) = 0
epoll_wait(4, {}, 32, 334)              = 0
stat("/etc/pgbouncer/userlist.txt", {st_mode=S_IFREG|0640, st_size=32, ...}) = 0
epoll_wait(4, {}, 32, 334)              = 0
stat("/etc/pgbouncer/userlist.txt", {st_mode=S_IFREG|0640, st_size=32, ...}) = 0
epoll_wait(4, {}, 32, 334)              = 0
stat("/etc/pgbouncer/userlist.txt", {st_mode=S_IFREG|0640, st_size=32, ...}) = 0
epoll_wait(4, {}, 32, 334)              = 0
stat("/etc/pgbouncer/userlist.txt", {st_mode=S_IFREG|0640, st_size=32, ...}) = 0
epoll_wait(4, {{EPOLLIN, {u32=13, u64=13}}}, 32, 334) = 1
accept(13, {sa_family=AF_LOCAL, NULL}, [2]) = 14
getsockopt(14, SOL_SOCKET, SO_PEERCRED, {pid=2226, uid=0, gid=0}, [12]) = 0
geteuid()                               = 276
fcntl(14, F_SETFD, FD_CLOEXEC)          = 0
fcntl(14, F_GETFL)                      = 0x2 (flags O_RDWR)
fcntl(14, F_SETFL, O_RDWR|O_NONBLOCK)   = 0
epoll_ctl(4, EPOLL_CTL_ADD, 14, {EPOLLIN, {u32=14, u64=14}}) = 0
write(2, "2015-11-12 18:23:08.452 2182 DEB"..., 1092015-11-12 18:23:08.452 2182 DEBUG C-0xbca520: (nodb)/(nouser)@unix(2226):6432 P: got connection: unix->unix
) = 109
write(3, "2015-11-12 18:23:08.452 2182 DEB"..., 109) = 109
accept(13, 0x7ffd577cde90, [2])         = -1 EAGAIN (Resource temporarily unavailable)
epoll_wait(4, {{EPOLLIN, {u32=14, u64=14}}}, 32, 57) = 1
recvfrom(14, "\0\0\0P\0\3\0\0user\0greg\0database\0postg"..., 2048, 0, NULL, NULL) = 80
write(2, "2015-11-12 18:23:08.452 2182 DEB"..., 982015-11-12 18:23:08.452 2182 DEBUG C-0xbca520: (nodb)/(nouser)@unix(2226):6432 got var: user=greg
) = 98
write(3, "2015-11-12 18:23:08.452 2182 DEB"..., 98) = 98
write(2, "2015-11-12 18:23:08.452 2182 DEB"..., 1062015-11-12 18:23:08.452 2182 DEBUG C-0xbca520: (nodb)/(nouser)@unix(2226):6432 got var: database=postgres
) = 106
write(3, "2015-11-12 18:23:08.452 2182 DEB"..., 106) = 106
write(2, "2015-11-12 18:23:08.453 2182 DEB"..., 1082015-11-12 18:23:08.453 2182 DEBUG C-0xbca520: (nodb)/(nouser)@unix(2226):6432 using application_name: psql
) = 108
write(3, "2015-11-12 18:23:08.453 2182 DEB"..., 108) = 108
write(2, "2015-11-12 18:23:08.453 2182 DEB"..., 1092015-11-12 18:23:08.453 2182 DEBUG C-0xbca520: (nodb)/(nouser)@unix(2226):6432 got var: client_encoding=UTF8
) = 109
write(3, "2015-11-12 18:23:08.453 2182 DEB"..., 109) = 109
write(2, "2015-11-12 18:23:08.453 2182 FAT"..., 1162015-11-12 18:23:08.453 2182 FATAL @src/objects.c:299 in function put_in_order(): put_in_order: found existing elem
) = 116
write(3, "2015-11-12 18:23:08.453 2182 FAT"..., 116) = 116
unlink("/var/run/pgbouncer/pgbouncer.pid") = 0
close(9)                                = 0
close(10)                               = 0
close(11)                               = 0
close(12)                               = 0
close(13)                               = 0
unlink("/tmp/.s.PGSQL.6432")            = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

**Comment by gclough** [comment]
I've verified that this still happens on v1.8.1:

**pgbouncer.ini**
```
[databases]
* = host=localhost port=5432 auth_user=pgbouncer

[pgbouncer]
auth_query = SELECT uname, phash FROM pgbouncer.user_lookup($1);
admin_users = postgres,gclough
```

**BUG TEST CASE**
```
[root@localhost pgbouncer]# cat userlist.txt 
"pgbouncer" "password"

[root@localhost pgbouncer]# /usr/pgsql-9.6/bin/psql -h 127.0.0.1 -U gclough -p 6432 postgres
Password for user gclough: 
psql (9.6.9)
Type "help" for help.

postgres=# \q

[root@localhost pgbouncer]# vi userlist.txt 
[root@localhost pgbouncer]# cat userlist.txt 
"pgbouncer" "password"
"gclough" "trustno1"

[root@localhost pgbouncer]# /usr/pgsql-9.6/bin/psql -h 127.0.0.1 -U gclough -p 6432 postgres
psql: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```

**Environment Details**
```
[root@localhost pgbouncer]# pgbouncer --version
pgbouncer version 1.8.1

[root@localhost pgbouncer]# cat /etc/redhat-release 
CentOS Linux release 7.5.1804 (Core)
```

**Comment by gclough** [comment]
**JUSTIFICATION**

This may seem like an obscure bug that nobody will ever hit, but consider the following circumstance:

1) User "X" logs into the database using `auth_query`, and queries some application data
2) They now need access to the `pgbouncer` database to query stats or manage pgbouncer... so they are added to the `userlist.txt`
3) At the next login pgbouncer crashes hard, and needs to be restarted

Yes, it's a corner case... but not a completely unimaginable case IMHO.

***WORKAROUND?***  To be perfectly honest, if I could find some way of NOT having to put `auth_users` into the `userlist.txt` file, then that would be Wonderful... but as yet, I've not found that to be possible.

**Comment by ali-joumma** [comment]
one workaround that worked for me is by using function instead of directly querying pg_shadow and  then make sure that the function doesn't return users specified in auth_file (by adding where statement: usename <> 'user in auth_file' )

**Comment by qris** [comment]
@ali-joumma It's not feasible to work around this problem, because of issue #59, where users listed in the pgbouncer.ini file must also be added to the `auth_user` file. So if you have an automatically-added user (via `auth_query`) and then you want to change any settings for them, you must list them in `pgbouncer.ini`, which (because of #59) means that you must add them to the `auth_file`, which kills pgBouncer when they next login.

Complete repro script:

```bash
#!/bin/sh

set -ex

# PGBOUNCER=~/external/src/pgbouncer/pgbouncer-1.10.0/pgbouncer
PGBOUNCER=../external/src/pgbouncer/pgbouncer/pgbouncer

shutdown_db_and_pgbouncer() {
    # Delete any existing Postgres database:
    if [ -d "db" ]; then
        if [ -r "db/postgresql.pid" ]; then
            pg_ctl -D db stop -m immediate
        fi
        rm -rf db
    fi

    if [ -r pgbouncer.pid ]; then
        kill -0 `cat pgbouncer.pid` && kill `cat pgbouncer.pid`
    fi
}

shutdown_db_and_pgbouncer

# Create and setup a new database:
initdb db
echo "external_pid_file = 'postgresql.pid'" >> db/postgresql.conf
pg_ctl -D db start

psql -h localhost postgres -c 'CREATE USER pgbadmin'
psql -h localhost postgres -c 'CREATE USER user1'
psql -h localhost -U pgbadmin -p 5432 postgres -c 'select 1'
psql -h localhost -U user1    -p 5432 postgres -c 'select 1'

# Setup initial conditions for the bug repro:
echo '"pgbadmin" ""' > pgbouncer_password.txt
[ -r "pgbouncer.pid" ] && kill `cat "pgbouncer.pid"` && sleep 1
rm -f pgbouncer.log

cat > pgbouncer.ini <<EOF
[pgbouncer]
listen_addr = *
listen_port = 7432
auth_type = trust
auth_file = pgbouncer_password.txt
auth_query = SELECT \$1, ''
logfile = pgbouncer.log
pidfile = pgbouncer.pid

[databases]
* = host=localhost port=5432 auth_user=pgbadmin
EOF

$PGBOUNCER -d pgbouncer.ini
while [ ! -r pgbouncer.pid ]; do sleep 0.1; done
read -p "Attach debugger to $PGBOUNCER pid $(cat pgbouncer.pid) and press Enter to continue"

# Start of the actual bug repro: connect as user1 to add them to pgBouncer using auth_query:
psql -h localhost -U user1 -p 7432 postgres -c 'select 1'

# Add user1 to the auth_user file, and reload config:
echo '"user1" ""' >> pgbouncer_password.txt
kill -HUP `cat pgbouncer.pid`
sleep 0.5

# Connecting as user1 will now kill pgBouncer:
psql -h localhost -U user1 -p 7432 postgres -c 'select 1' || true

shutdown_db_and_pgbouncer

# Make sure that it actually failed as expected:
grep "FATAL @src/objects.c:.* in function put_in_order(): put_in_order: found existing elem" pgbouncer.log || echo "Repro failed"

exit 0
```

**Comment by zurikus** [comment]
Hello to everyone, 
hit this bug on our prod environment today.
As I can see there is the fix provided by @qris. Is there any reason why it is not merged into master?

**Comment by JelteF** [comment]
Closing this since @eulerto reported that this was fixed at least in 1.17 already.
